torch/autograd/gradcheck.py

```diff
@@ -1656,42 +1656,51 @@ def _to_real_dtype(dtype):
-        if x.layout in {torch.sparse_csr, torch.sparse_bsr}:
-            compressed_indices, plain_indices = x.crow_indices(), x.col_indices()
+    # complex tensor with only real component. Tensors are default allocated on
+    # the CPU device, as the generator is located there too.
+    with torch.device("cpu"):
+        if x.layout == torch.sparse_coo:
+            # For sparse, create a random sparse vec with random values in the same
+            # indices. Make sure size is set so that it isn't inferred to be smaller.
+            x_values = x._values()
+            dtype = _to_real_dtype(x.dtype) if downcast_complex else x.dtype
+            values = (
+                torch.rand(x_values.numel(), generator=generator)
+                .to(dtype=dtype, device=x.device)
+                .view(x_values.shape)
+            )
+            values /= values.norm()
+            vec = torch.sparse_coo_tensor(
+                x._indices(), values, x.size(), device=x.device
+            )
+        elif _is_sparse_compressed_tensor(x):
+            if x.layout in {torch.sparse_csr, torch.sparse_bsr}:
+                compressed_indices, plain_indices = x.crow_indices(), x.col_indices()
+            else:
+                compressed_indices, plain_indices = x.ccol_indices(), x.row_indices()
+            x_values = x.values()
+            dtype = _to_real_dtype(x.dtype) if downcast_complex else x.dtype
+            values = (
+                torch.rand(x_values.numel(), generator=generator)
+                .to(dtype=dtype, device=x.device)
+                .view(x_values.shape)
+            )
+            values /= values.norm()
+            vec = torch.sparse_compressed_tensor(
+                compressed_indices,
+                plain_indices,
+                values,
+                x.size(),
+                layout=x.layout,
+                device=x.device,
+            )
         else:
-            compressed_indices, plain_indices = x.ccol_indices(), x.row_indices()
-        x_values = x.values()
-        dtype = _to_real_dtype(x.dtype) if downcast_complex else x.dtype
-        values = (
-            torch.rand(x_values.numel(), generator=generator)
-            .to(dtype=dtype, device=x.device)
-            .view(x_values.shape)
-        )
-        values /= values.norm()
-        vec = torch.sparse_compressed_tensor(
-            compressed_indices, plain_indices, values, x.size(), layout=x.layout
-        )
-    else:
-        dtype = _to_real_dtype(x.dtype) if downcast_complex else x.dtype
-        vec = torch.rand(x.numel(), generator=generator).to(
-            dtype=dtype, device=x.device
-        )
-        vec /= vec.norm()
-    return vec
+            dtype = _to_real_dtype(x.dtype) if downcast_complex else x.dtype
+            vec = torch.rand(x.numel(), generator=generator).to(
+                dtype=dtype, device=x.device
+            )
+            vec /= vec.norm()
+        return vec
```

Comment:
In theory `_vec_from_tensor` is agnostic to generator types. I'd prefer if we'd update:
```python
def _make_vectors(inp_tensors, outputs, *, use_forward_ad):
    # Use our own generator to avoid messing with the user's RNG state
    g_cpu = torch.Generator()

    def _vec_from_tensor_cpu(*args)
         with torch.device("cpu"):
              return _vec_from_tensor(*args)
    all_u = []
    all_u_dense = []
    for inp in inp_tensors:
        ur = _vec_from_tensor_cpu(inp, g_cpu, True)
        ur_dense = _to_flat_dense_if_sparse(ur)
        if inp.is_complex():
            ui = _vec_from_tensor_cpu(inp, g_cpu, True)
            all_u.append((ur, ui))
            ui_dense = _to_flat_dense_if_sparse(ui)
            all_u_dense.append((ur_dense, ui_dense))
        else:
            all_u.append(ur)
            all_u_dense.append(ur_dense)
    all_v = (
        None if use_forward_ad else [_vec_from_tensor_cpu(out, g_cpu) for out in outputs]
    )
    return all_v, all_u, all_u_dense
    ```